### PR TITLE
Pool buffers used in S3.WriteStream

### PR DESCRIPTION
**Pull request by stevvooe**
Pooling these buffers should help to reduce garbage collection pressure experience when using the s3 driver. The theory is that these are causing timeouts while sending requests to s3's backend.

Signed-off-by: Stephen J Day <stephen.day@docker.com>

### CODE REVIEW

**Comment by stevvooe** [comment]
cc @dmp42 @dmcgowan @AndreyKostov

File: registry/storage/driver/s3/s3.go [hunk 1]
```diff
@@ -314,6 +321,9 @@ func (d *driver) WriteStream(path string, offset int64, reader io.Reader) (total
 				}
 			}
 		}
+
+		d.putbuf(buf)
```
Comment (dmcgowan) [inline registry/storage/driver/s3/s3.go]: This doesn't match up with the existing logic, currently this block never reallocs buf

File: registry/storage/driver/s3/s3.go [hunk 2]
```diff
@@ -287,8 +294,8 @@ func (d *driver) WriteStream(path string, offset int64, reader io.Reader) (total
 		return 0, err
 	}
 
-	buf := make([]byte, d.ChunkSize)
-	zeroBuf := make([]byte, d.ChunkSize)
+	buf := d.getbuf()
+	zeroBuf := d.getbuf()
```
Comment (dmcgowan) [inline registry/storage/driver/s3/s3.go]: Since zeroBuf is only used when offset >0, can this getBuf and putBuf just be put at the boundaries of that block?

**Comment by RichardScothern** [comment]
LGTM

File: registry/storage/driver/s3/s3.go (hunk 1 above)
Comment (stevvooe) [inline registry/storage/driver/s3/s3.go]: This defer is run at the end of the function. I guess these can be outside this defer.

File: registry/storage/driver/s3/s3.go [hunk 3]
```diff
@@ -724,3 +736,17 @@ func getPermissions() s3.ACL {
...
+
+// getbuf returns a buffer from the driver's pool with length d.ChunkSize. The
+// buffer provided is cleared of any data.
+func (d *driver) getbuf() []byte {
+	return d.pool.Get().([]byte)
+}
+
+func (d *driver) putbuf(p []byte) {
+	// clear before returning buffer
+	for i := 0; i < len(p); i++ {
```
Comment (dmcgowan) [inline registry/storage/driver/s3/s3.go]: Why is this necessary, the buffer is always explicitly overwritten by read commands and the read length is checked.

**Comment by dmcgowan** [comment]
LGTM

File: registry/storage/driver/s3/s3.go [hunk 4]
```diff
@@ -367,6 +376,8 @@ func (d *driver) WriteStream(path string, offset int64, reader io.Reader) (total
 		}
 
 		go func(bytesRead int, from int64, buf []byte) {
+			defer d.putbuf(buf) // this buffer gets dropped after this call
```
Comment (endophage) [inline registry/storage/driver/s3/s3.go]: We're going to end up putting the same ```buf``` twice. The ```d.putbuf(buf)``` further up will be bound to the same instance of ```buf```

Comment (stevvooe) [inline registry/storage/driver/s3/s3.go]: @endophage Thanks for the keen eye!

**Comment by dmcgowan** [comment]
LGTM

**Comment by endophage** [comment]
LGTM
